source/local/local.go

```diff
@@ -144,17 +160,54 @@ func (s *localSource) GetFeatures() *nfdv1alpha1.Features {
 	return s.features
 }
 
-func parseFeatures(lines [][]byte) map[string]string {
+func parseDirectives(line string, fileName string, opts *parsingOpts) error {
+	lineSplit := strings.SplitN(line, "=", 2)
+
+	// If we have the following directive: # +expiry-time=xxx
+	// we need to get the key '# +expiry-time' and then we need to extract
+	// the following key 'expiry-time'.
+	key := strings.Split(lineSplit[0], DirectivePrefix)[1]
+	value := lineSplit[1]
+
+	switch key {
+	case ExpiryTimeKey:
+		expiryDate, err := time.Parse(time.RFC3339, strings.TrimSpace(value))
+		if err != nil {
+			klog.ErrorS(err, "failed to parse feature file expiry date", "fileName", fileName)
+			return err
+		}
+		opts.ExpiryTime = expiryDate
+	default:
+		klog.ErrorS(ErrorUnknownDirective, "unknown directive", "directiveKey", key)
+		return ErrorUnknownDirective
+	}
+
+	return nil
+}
+
+func parseFeatures(lines [][]byte, fileName string) map[string]string {
 	features := make(map[string]string)
+	now := time.Now()
+	parsingOpts := &parsingOpts{
+		ExpiryTime: now,
+	}
 
 	for _, l := range lines {
 		line := strings.TrimSpace(string(l))
 		if len(line) > 0 {
-			// Skip comment lines
-			if strings.HasPrefix(line, "#") {
+			if strings.HasPrefix(line, DirectivePrefix) {
+				// Parse directives
+				err := parseDirectives(line, fileName, parsingOpts)
+				if err != nil {
+					klog.ErrorS(err, "error while parsing directives", "fileName", fileName)
+				}
+
+				// handle expiration
+				if parsingOpts.ExpiryTime.Before(now) {
+					return map[string]string{}
+				}
```

Comment:
For this feature, we are only supporting one `expiry-time` for the whole file. it's not customizable for each label. Maybe if there's a need for that kind of customization we can adjust. wdyt @marquiz @eero-t?

Reply:
Making it more flexible would be just one line change (+ doc update).   Otherwise I think code should return error if there are features specified before directives (+ doc change).

Reply:
Thanks for review @eero-t . Now that I look at this with fresher eyes after a break I think @eero-t's comment/suggestion above makes sense.

I feel that the logical thing to do would be to treat this like batch processing/shell scripts: read and execute line by line. The directive takes effect when it is encountered, think about bash `set +x ... set -x for example. That is, you could have different expiry times for different features as outlined above). If would be directives that are for the whole file, they should be placed at the very top of the file (like go build tags or bash shebang or similar). I don't think we need to have this limitation for the `expiry-time` directive.

WDYT @AhmedGrati @eero-t

Reply:
To me that makes sense.

Doc would then need to be changed to something like this:
> `expiry-time` option defines date/time when all features listed after it will expire / are ignored:
> ```
> # +expiry-time=2023-12-31
> yearly_label=value1
> 
> # +expiry-time=2023-08-31
> monthly_label=value2
> ```
> (Meaning that any `expiry-time` having later date/time than previous ones, will have no effect.)

PS. If support for directives that can apply only to whole file is added later on, it's more consistent for the user if their implementation requires that they are set before any of the features are specified.

Reply:
> If support for directives that can apply only to whole file is added later on, it's more consistent for the user if their implementation requires that they are set before any of the features are specified.

Yes

Reply:
@marquiz how should we know what are the specific features that a directive would be applied on? e.g.: Should we apply the expiry directive on all features until we encounter the same directive key, or we should have a delimiter like `\n`? 

Also, we should consider the case where we can add more directives, how should we handle these batches? e.g.:
```
# +expiry-time=2023-12-31
# +key1=value1
yearly_label=value1

# +expiry-time=2023-08-31
monthly_label=value2
```

would `key1` be applied only on `yearly_label`, or on both `yearly_label` and `monthly_label`.

Reply:
It depends at which point in the file NFD encounters expiration.

If it's 2024, none of them is applied, because file processing expires on first line.

If its late 2023, processing expires on the `+expiry-time=2023-08-31` line.

If it's early 2023, end of that file is reached without file processing expiring.

This is all you need for that (besides doc update):
```suggestion
				if parsingOpts.ExpiryTime.Before(now) {
					return feature
				}
```

Reply:
To make it stupid and simple, I think those directives could/should be treated like "variables". Expiry-time (or whatever key) is applied to all features until a new expiry time is specified. WDYT?

Reply:
@marquiz I got your point. The question is, when we encounter an expired feature, should we continue parsing the rest of the file, or we just return features that we already parsed and not expired?

Reply:
I think the logical thing to do would be to always process the whole file